### ISSUE TITLE
Training in 4-Agent Obstacle Scenario

### ISSUE THREAD

**Issue by YouMinJung** [Zhehui-Huang/quad-swarm-rl#97]
Hello,
I am conducting experiments to reproduce the results presented in the paper using this repository. All experiments are being performed within a Docker-based containerized environment (Ubuntu 22.04, Python 3.11).

Despite following the general guidelines, I am unable to achieve satisfactory performance in the 4-agent swarm with obstacles scenario.
I have completed approximately 12.8% of the total planned training. Although it is still in the relatively early stages of the full training curriculum, the following trends are concerning.

- The Min Reward initially shows signs of improvement but eventually   plateaus or gradually declines as training progresses.
- While the success rate exhibits a marginal upward trend, it remains critically low (near-zero).

**Experimental Setup & Logs**

**Command Line**: 
```
python -m swarm_rl.train --env=quadrotor_multi --quads_num_agents=4 --quads_mode=static_same_goal \
--with_wandb=False --device=gpu --serial_mode=True \
--train_for_env_steps=1000000000 --algo=APPO --use_rnn=False \
--num_workers=4 --num_envs_per_worker=4 --learning_rate=0.0001 --ppo_clip_value=5.0 --recurrence=1 \
--nonlinearity=tanh --actor_critic_share_weights=False --policy_initialization=xavier_uniform \
--adaptive_stddev=False --with_vtrace=False --max_policy_lag=100000000 --rnn_size=256 \
--gae_lambda=1.00 --max_grad_norm=5.0 --exploration_loss_coeff=0.0 --rollout=128 --batch_size=1024 \
--with_pbt=False --normalize_input=False --normalize_returns=False --reward_clip=10 \
--quads_use_numba=True --save_milestones_sec=3600 --anneal_collision_steps=300000000 \
--replay_buffer_sample_prob=0.75 \
--quads_episode_duration=15.0 \
--quads_obs_repr=xyz_vxyz_R_omega_floor \
--quads_neighbor_hidden_size=256 --quads_neighbor_obs_type=pos_vel --quads_collision_hitbox_radius=2.0 \
--quads_collision_falloff_radius=4.0 --quads_collision_reward=5.0 --quads_collision_smooth_max_penalty=4.0 \
--quads_neighbor_encoder_type=no_encoder --quads_neighbor_visible_num=2 \
--quads_use_obstacles=True --quads_obst_spawn_area 8 8 --quads_obst_density=0.2 --quads_obst_size=0.6 \
--quads_obst_collision_reward=5.0 --quads_obstacle_obs_type=octomap --quads_use_downwash=True \
--experiment=test_obst_avoid
```

Running on Docker. When increasing num_workers for better throughput, I observed signal_slot synchronization issues, so the current results are obtained from a stabilized configuration (e.g., lower worker count or serial mode).
I have attached the TensorBoard screenshots demonstrating the reward/success rate degradation.

![Image](https://github.com/user-attachments/assets/cd257672-beea-4aaa-b7c2-35f16a5b356f)
![Image](https://github.com/user-attachments/assets/1b1edcf7-677b-47c6-abf7-06402cee3f10)

​I would appreciate your insights on the following points to help me troubleshoot the lack of convergence:

- Due to synchronization issues in my Docker environment, I have been conducting experiments with serial_mode set to True. While I understand this significantly reduces throughput, could this setting qualitatively affect the training stability or the gradient updates in a way that prevents convergence for complex tasks like obstacle avoidance? Is there any known discrepancy in learning behavior between serial and parallel/asynchronous modes?
- ​Given that the rewards are declining and the success rate is not improving for the 4-agent swarm with obstacles, what would be the recommended next steps?

​I am eager to hear your suggestions on how to modify my setup or parameters to achieve the performance reported in your work.
